[TRTLLM-9735][feat] Add processed logprobs functionality to TorchSampler

stnie opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added support for configurable log probability modes ("raw" or "processed") to control logprob output format.
  * Enhanced per-token sampled log probability tracking with top-k support.
  * Improved sampling strategy handling with temperature state management.

* **Bug Fixes**
  * Fixed log probability parameter handling to respect explicit false values.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

- Iintroduces a new optional parameter, logprobs_mode, to the SamplingParams and LlmRequest classes, allowing users to specify the mode of log probabilities to return.
- Create process_logprobs function to remove logprobs processing code from process_requests.
- add batching based on logprobs_mode to sample_batched_by_strategy
- additionally return processed logits from sampling

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

hchings committed:
sampled logprob

fix step

remove dependency on return_genereation_logits

align API across backends, add tests

test

fix group sampling strategy

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>
Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

hchings committed:
update

Signed-off-by: Erin Ho <14718778+hchings@users.noreply.github.com>
Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

stnie committed:
[None][feat] Add processed logprobs functionality to TorchSampler

- Iintroduces a new optional parameter, logprobs_mode, to the SamplingParams and LlmRequest classes, allowing users to specify the mode of log probabilities to return.
- Create process_logprobs function to remove logprobs processing code from process_requests.
- add batching based on logprobs_mode to sample_batched_by_strategy
- additionally return processed logits from sampling

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

stnie committed:
[TRTLLM-9686][feat] Fix issues with processed logprobs functionality.

- Expand test_logits_logprobs to perform a check for processed logprobs
- Fix processed logprobs for greedy sampling and when using temperature

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

hchings commented on tensorrt_llm/sampling_params.py:
```diff
@@ -321,6 +324,11 @@ def _validate(self):
                 f"under the greedy decoding."
             )
 
+        if self.logprobs_mode not in ["raw", "processed"]:
```
nit: since this flag is for PyTorch backend only, it'd be better that we also add a guard for it.

stnie committed:
[TRTLLM-9687][fix] Enable pinned memory for tensor allocations in TorchSampler

- Updated tensor allocation in TorchSampler to use pinned memory for improved performance during D2H copies.
- Modified test_sampled_token_always_in_logprobs to include logprobs_mode parameter for enhanced testing of log probabilities.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

ixlmar reviewed (commented):
Reviewed code minus tests.

Re the type annotations, I recommend checking using some type checker (e.g. in the IDE) as long as we don't have that in the CI. Ideally, we could maintain type correctness for `sampling_utils.py`, `flashinfer_sampling_utils.py`, and the newer parts of `sampler.py`.

stnie committed:
[TRTLLM-9688][feat] Refactor processed logprobs to skip logprob calculation when not needed

- Added LogProbsMode class to define modes for log probabilities: RAW and PROCESSED.
- Updated SamplingParams and LlmRequest to utilize LogProbsMode for logprobs_mode parameter.
- Enhanced validation to check logprobs_mode against LogProbsMode values.
- Modified TorchSampler and related classes to support new logprobs_mode functionality.
- Modified TorchSampler to only calculate logprobs when a request needs it
- Updated tests to cover new logprobs_mode behavior and ensure correct processing of log probabilities.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #27528](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/27528/) [ run ] triggered by Bot. Commit: `18531f1`

tensorrt-cicd commented:
[PR_Github #27528](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/27528/) [ run ] completed with state `SUCCESS`. Commit: `18531f1` 
[/LLM/main/L0_MergeRequest_PR pipeline #21006](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/21006/) completed with status: 'FAILURE'

tongyuantongyu commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -2487,6 +2683,95 @@ def request_stop_words(request: LlmRequest, new_tokens: torch.Tensor):
+            req = requests[req_id]
+            next_offset = current_offset + steps
+            # NB: Assigning views on memory which is being filled asynchronously
+            req.py_topk_logprobs_vals = topk_vals[current_offset:next_offset, : req.py_num_logprobs]
```
Creating these views can be slow. We should do `{topk,sampled}_{vals,indices,rank}.tolist()` and extract data for each request from the python list.

ixlmar replied:
In this particular case for `.py_topk_logprobs_vals`, there is only one PyTorch bindings invocation per request. Since the memory referenced by those views is not yet filled, `.tolist` cannot be used without a refactor.

After staring at this for a while, I think `.py_topk_logprobs_vals` is actually only accessed in `handle_logprobs`. So, instead of attaching (for each request), tensor slices to `LlmRequest` and then converting them to native Python data types in `_convert_logprobs_tensor_to_list`, we could:
* Include the full `topk_vals` tensor in the sampler state (keeping the async D2H transfer)
* Call `topk_vals.tolist()` in `handle_logprobs` once and process the data.

This approach should reduce the binding overheads from O(num_requests) to O(1).

(If we keep the current implementation, we should `del req.py_topk_logprobs_vals` in `handle_logprobs()` to release the tensor it references.)

stnie replied:
I refactored the logprobs processing to be more performant

stnie committed:
[TRTLLM-9689][feat] Introduce max_topk_logprobs parameter and enhance logprobs handling

- Added max_topk_logprobs parameter to AutoDeployConfig and LlmRequest to control the number of top-k logprobs storable for each token.
- Updated TorchSampler to accommodate max_topk_logprobs in logprobs processing and storage.
- Enhanced logprobs handling in the sampling process to support both sampled and top-k logprobs.
- Enabled batched processing of logprobs to enhance logprobs performance
- Modified tests to validate the new max_topk_logprobs functionality and ensure correct logprobs output.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>
Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #28059](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28059/) [ run ] triggered by Bot. Commit: `5f75078`

stnie commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #28391](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28391/) [ run ] triggered by Bot. Commit: `ad2b24e`

tongyuantongyu committed:
cache sampling_strategy

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

tongyuantongyu committed:
fix apply_embedding_bias perf

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

tongyuantongyu committed:
fix sample_batched_by_strategy perf

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

stnie committed:
[TRTLLM-9686][chore] Added skip to testcase and fixed _request_strategy when using beam search

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #28607](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28607/) [ run ] triggered by Bot. Commit: `1def846`

tensorrt-cicd commented:
[PR_Github #28607](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28607/) [ run ] completed with state `SUCCESS`. Commit: `1def846` 
[/LLM/main/L0_MergeRequest_PR pipeline #21922](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/21922/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

stnie committed:
[TRTLLM-9686][chore] Refactored speculation_needs_probs to use request indices instead of a boolean tensor

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run --disable-fail-fast

stnie commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -338,13 +339,19 @@ def _group_requests_by_strategy_key(
-            strategy=group_key[0], speculation_needs_probs=group_key[1]
+            strategy=group_key[0],
+            speculation_needs_probs=group_key[1],
+            need_processed_logprobs=group_key[2],
```
I refactored this

tensorrt-cicd commented:
[PR_Github #28776](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28776/) [ run ] completed with state `SUCCESS`. Commit: `6b2950d` 
[/LLM/main/L0_MergeRequest_PR pipeline #22023](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/22023/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #28809](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28809/) [ run ] triggered by Bot. Commit: `6b2950d`

tensorrt-cicd commented:
[PR_Github #28809](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28809/) [ run ] completed with state `SUCCESS`. Commit: `6b2950d` 
[/LLM/main/L0_MergeRequest_PR pipeline #22054](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/22054/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #28939](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28939/) [ run ] triggered by Bot. Commit: `6b2950d`

tensorrt-cicd commented:
[PR_Github #28939](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/28939/) [ run ] completed with state `SUCCESS`. Commit: `6b2950d` 
[/LLM/main/L0_MergeRequest_PR pipeline #22168](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/22168/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tongyuantongyu commented on tests/unittest/_torch/sampler/test_logits_logprobs.py:
```diff
@@ -0,0 +1,605 @@
+                assert logprob_diff < 1e-4, \
+                    f"Req {req_idx}, Token {token_idx}: Logprob mismatch! " \
+                    f"Returned {returned_logprob:.6f} but expected {expected_logprob:.6f} " \
+                    f"(diff={logprob_diff:.6f})."
```
Consider [`torch.testing.assert_close`](https://docs.pytorch.org/docs/stable/testing.html#torch.testing.assert_close) instead of manual comparing.

stnie committed:
Refactor logprob assertions in test_processed_logprobs_e2e to use torch.testing.assert_close for improved readability and precision

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tongyuantongyu committed:
Nits, Update to ruff for test file

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

Superjomn reviewed (approved):
LGTM on the llmapi changes.

zhenhuaw-me commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -566,6 +568,8 @@ def __init__(
         # currently, keep py_stop_words_list as python list, rather than tensor.
         self.py_stop_words_list = stop_words_list
 
+        self.py_logprobs_mode = LogprobMode.RAW if logprobs_mode is None else logprobs_mode
```
Does it make sense to let the parameter default to `LogprobMode.RAW` explicitly?

litaotju replied:
If so, can we make the default as `RAW` instead of None in the function declaration?

yweng0828 commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -862,31 +955,54 @@ def __post_init__(self):
+        finish_reasons = int_tensor(self.NEW_TOKENS_SHAPE)
+
+        # Only used for logprobs processing or beam search
+        sampled_log_probs = torch.zeros(self.LOGPROBS_SHAPE, device="cuda", dtype=torch.float32)
```
Can we just use `tensor.empty()` here? IIUC, `torch.zeros()` may introduce additional memset overhead.

tongyuantongyu committed:
Update logprobs_mode definition

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

tongyuantongyu committed:
Avoid unnecessary initialization for runtime buffers

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

litaotju commented on tensorrt_llm/sampling_params.py:
```diff
@@ -329,8 +347,11 @@ def _validate(self):
+        if self.logprobs is False:
+            self.logprobs = None
+        if self.logprobs is True:
+            self.logprobs = 0
```
I don't understand whats the different meaning of None and 0 here?

Can be just use a typed bool? instead of mixing boolean/None/0-1?

litaotju commented:
Could you pls run some performance tests and make sure there is no addtional sync in ths sampler part? Which can easily introduce regressions based on recent changes on sampler.

litaotju commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -1836,6 +1836,8 @@ def _validate_request(self, request: LlmRequest):
                     f"is not equal to max_beam_width {self.max_beam_width}. This is not supported!"
                 )
+        # Validate logprobs mode
+        request.validate_logprobs_mode()
```
shall we do this validation in the Request post init function?

litaotju commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -593,6 +597,15 @@ def set_exclude_last_generation_logits(
 
+    def validate_logprobs_mode(self):
+        if self.py_logprobs_mode not in [
+                LogprobMode.RAW, LogprobMode.PROCESSED
```
If None is never allowed, is line 489 is still correct?

litaotju commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -88,6 +88,34 @@
+
+@dataclass(kw_only=True)
+class LogProbsStateList:
+    sampled_vals: list[list[list[float]]]
```
Many duplicates of this nested `list[list[list` type, shall we consider using https://docs.python.org/3/library/typing.html#type-aliases ?

litaotju commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -358,8 +406,10 @@ def _request_get_sampling_params(request: LlmRequest) -> UtilsSamplingParams:
-    return resolve_sampling_strategy(params, vocab_size=vocab_size)
+    if not hasattr(request, "py_sampling_strategy") or _get_max_beam_width(request) > 1:
+        params = _request_get_sampling_params(request)
+        request.py_sampling_strategy = resolve_sampling_strategy(params, vocab_size=vocab_size)
```
I noticed this function changes its previous behavior that don't modify the input request to now modifying it.

Is this really what we want here?

hchings replied:
I think this is to avoid repeated computation by only compute and store on request.py_sampling_strategy once, and reuse the cached value after.

litaotju commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -266,7 +266,8 @@ def greedy_search_sampling_batch(
     if return_probs:
-        softmax = torch.softmax(logits, dim=-1)
+        softmax = torch.zeros_like(logits)
+        softmax.scatter_(1, next_tokens.unsqueeze(-1), 1.0)
```
I think this changes the semantics, when applying softmax, the probs for other tokens not being select is the value of softmax function or just 0?

How does other FW do?

litaotju commented on tensorrt_llm/_torch/pyexecutor/sampling_utils.py:
```diff
@@ -583,7 +585,7 @@ def sample_grouped_strategies(
         return_probs: bool,
         group_metadata: StrategyMetadata | None = None,
-    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor], float | None]:
```
When shall we use `Optional[float]` and when to use `float | None`?

zhenhuaw-me replied:
I think it's like "west const" vs "east const" - it's a convention issue...

tongyuantongyu replied:
As explained before I think we are moving to the new grammar but don't want to suddenly change everything.

hchings reviewed (approved):
@tongyuantongyu  Could you share some benchmark or Nsight screenshots of the `_process_logprobs` in the MR description as well? Thanks.

zhenhuaw-me commented on .pre-commit-config.yaml:
```diff
@@ -1022,7 +1022,6 @@ common-files: &common_files |
         tests/unittest/_torch/ray_orchestrator/single_gpu/test_cache_transceiver_comm.py |
         tests/unittest/_torch/sampler/test_beam_search.py |
         tests/unittest/_torch/sampler/test_best_of_n.py |
-        tests/unittest/_torch/sampler/test_return_logits.py |
```
Should we add `tests/unittest/_torch/sampler/test_logits_logprobs.py`?

tongyuantongyu commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -819,6 +899,7 @@ def _record_sampler_event(self) -> SamplerEvent:
 
 class TorchSampler(Sampler, AsyncWorkerMixin):
     SampleState = SampleStateTorch
+    DEFAULT_MAX_TOPK_LOGPROBS = 20
```
That affects the default size of `topk_indices` and `topk_vals`. They are initialized with size `(max_num_sequences, max_draft_tokens + 1, DEFAULT_MAX_TOPK_LOGPROBS)` which is unlikely to go beyond several MBs.

So if 20 is a too large estimation the overhead is negligible. If it's too small, then there will be a one-time expansion and it also shouldn't affect perf too much.

I don't think it worth a perf knob to be considered by the user for tuning.

tongyuantongyu commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -88,6 +88,34 @@
+    @staticmethod
+    def from_logprobs_state(logprobs_state: LogProbsState) -> "LogProbsStateList":
+        return LogProbsStateList(
+            sampled_vals=logprobs_state.sampled_vals.tolist(),
```
They are host tensors. We are specially doing this to avoid binding call overhead when indexing tensors.

zhenhuaw-me commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -485,6 +486,7 @@ def __init__(
             is_first_draft: bool = False,
             use_chunked_generation_logits: bool = True,
             logits_chunk_size: int = 8,
+            logprobs_mode: LogprobMode | None = None,
```
IIUC, we don't allow `logprobs_mode == None` in our codebase right? It defaults to `RAW` at API level and is passing around with a "real value".

tongyuantongyu replied:
Updated to `logprobs_mode: LogprobMode = LogprobMode.RAW`.

tongyuantongyu commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -1820,12 +2025,26 @@ def update_requests(
     def _return_log_probs(self, requests: list[LlmRequest]) -> bool:
         return any(req.py_return_log_probs for req in requests)
 
+    def _prepare_log_probs(self, requests: list[LlmRequest]) -> None:
```
If we write nothing, then it's up to IDE to infer the actual return type. Documentation shows examples that explicitly types `-> None`: https://docs.python.org/3/library/typing.html#type-aliases

zhenhuaw-me commented on tensorrt_llm/sampling_params.py:
```diff
@@ -219,6 +234,7 @@ class SamplingParams:
     n: int = 1
     best_of: Optional[int] = None
     use_beam_search: bool = False
+    logprobs_mode: LogprobMode = LogprobMode.RAW
```
Better to move close to other logprobs attributes?

zhenhuaw-me commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -370,9 +420,10 @@ def _group_requests_by_strategy_key(
-    )
+    group_dict: dict[
+        tuple[GenericStrategyKeyType, bool],
+        tuple[list[int], list[Strategy], list[int], list[bool], list[bool]],
```
This is pretty hard to read. Would it be nicer to define a alias type for it?

tongyuantongyu committed:
Address reviews

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

tongyuantongyu commented on tensorrt_llm/sampling_params.py:
```
# Keep the below fields in sync with tllme.OutputConfig
```

We don't want to add it to the legacy binding type `tllme.OutputConfig`.

zhenhuaw-me commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -88,6 +88,37 @@
+@dataclass(kw_only=True)
+class LogProbsStateList:
+    FloatState = list[list[list[float]]]
+    IntState = list[list[list[float]]]
```
```suggestion
    IntState = list[list[list[int]]]
```
Typo?

tongyuantongyu committed:
Address reviews 2

Signed-off-by: Yuan Tong <13075180+tongyuantongyu@users.noreply.github.com>

tongyuantongyu commented:
/bot run

tensorrt-cicd commented:
[PR_Github #31220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31220/) [ run ] triggered by Bot. Commit: `dcf0a9d`

tensorrt-cicd commented:
[PR_Github #31220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31220/) [ run ] completed with state `SUCCESS`. Commit: `dcf0a9d` 
[/LLM/main/L0_MergeRequest_PR pipeline #24127](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24127/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

stnie committed:
Fix annotations for api_stability tests

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #31533](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31533/) [ run ] triggered by Bot. Commit: `28c33cf`

stnie committed:
Added testcases to waives.txt to continue skipping these testcases after renaming test_return_logits.py to test_logits_logprobs.py

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #31996](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31996/) [ run ] triggered by Bot. Commit: `8c6bea1`

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #32092](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32092/) [ run ] triggered by Bot. Commit: `8c6bea1`

hchings commented on tensorrt_llm/executor/result.py:
```diff
@@ -933,6 +933,21 @@ def _topk_logprobs(logits: torch.Tensor, top_k: int,
+            if tokens is not None:
+                for t in range(logprobs.size(0)):
+                    token_id = tokens[t]
+                    token_logprob = logprobs[t, token_id].item()
```
Yes this is for TRT flow and we can leave it as it'll be deprecated.
